src/sdk/namespace_fs.js

```diff
@@ -593,21 +598,21 @@ class NamespaceFS {
                 const source_file_path = path.join(this.bucket_path, params.copy_source.key);
                 await this._fail_if_archived_or_sparse_file(fs_account_config, source_file_path);
                 try {
-                    const same_inode = await this._is_same_inode(fs_account_config, source_file_path, file_path);
+                    const same_inode = await this._is_same_inode(fs_account_config, source_file_path, file_path, fs_xattr);
                     if (same_inode) return same_inode;
                     // Doing a hard link.
                     await nb_native().fs.link(fs_account_config, source_file_path, upload_path);
                 } catch (e) {
                     dbg.warn('NamespaceFS: COPY using link failed with:', e);
-                    await this._copy_stream(source_file_path, upload_path, fs_account_config, fs_xattr);
+                    fs_xattr = await this._copy_stream(source_file_path, upload_path, fs_account_config, fs_xattr);
                 }
             } else {
-                await this._upload_stream(params.source_stream, upload_path, fs_account_config, object_sdk.rpc_client, fs_xattr);
+                fs_xattr = await this._upload_stream(params.source_stream, upload_path, fs_account_config, object_sdk.rpc_client, fs_xattr);
             }
             // TODO use file xattr to store md5_b64 xattr, etc.
             const stat = await nb_native().fs.stat(fs_account_config, upload_path);
```

Comment:
Similar to @guymguym Comments,
If we are returning `fs_xattr` from `_copy_stream` and `_upload_stream` and we have opened the file there, we should `stat` the fd (using `fstat`) and return it also, then skip this `stat`.